rfc/router-storage.md

```diff
@@ -0,0 +1,767 @@
+
+Handling of `blob/accept` on the router is as follows:
+
+1. Lookup Storage node that received data
+
+```js
+// psuedoCode
+const blobAllocation = blobAllocations[acceptInvocation.args.blob.digest.bytes]
+const storageNode = storageNodes[blobAllocation.did]
+```
+
+2. Invoke `blob/accept` on storage node
+
+```json
+ // 3. System will attempt to accept uploaded content that matches blob
+  // multihash and size.
+  { // "/": "bafy...accept",
+
+    "iss": "did:web:web3.storage",
+    "aud": "did:web:node1.storage",
+    "cmd": "/service/blob/accept",
+    "sub": "did:web:web3.storage",
+    "args": {
+      // as best I can tell this is used to determine who is the audience for the location commitment, which should be web3.storage
+      "space": "did:web:web3.storage",
+      "blob": {
+        // multihash of the blob as byte array
+        "content": { "/": { "bytes": "mEi...sfKg" } },
+        "size": 2_097_152,
+      },
+      "expires": 1712903125,
+    }
+  }
+```
+
+Legacy Invocation:
+
+```json
+{ // "/": "bafy...accept",
+  "iss": "did:web:web3.storage",
+  "aud": "did:web:node1.storage",
+  "att": {
+    "can": "/web3.storage/blob/accept",
+    "with": "did:web:web3.storage",
+    "nb": {
+      // as best I can tell this is used to determine who is the audience for the location commitment, which should be web3.storage?
+      // alternatively maybe it should be directly to space
+      "space": "did:web:web3.storage",
+      "blob": {
+        // multihash of the blob as byte array
+        "content": { "/": { "bytes": "mEi...sfKg" } },
+        "size": 2_097_152,
+      },
+      "expires": 1712903125,
+    }
+  }
+}
+```
+
+Receipt:
+
+```json
+{ 
+  // "/": "bafy..work",
+  "iss": "did:web:node1.storage",
+  "aud": "did:web:web3.storage",
+  "cmd": "/ucan/assert",
+  "sub": "did:web:web3.storage",
+  "args": {
+    "assert": [
+      // refers to the invocation from the example
+      { "/": "bafy..accept" },
+      // refers to the receipt corresponding to the above invocation
+      { 
+        "out": {
+          "ok": {
+            // result is commitment that is produced as result of "bafy..accept"
+            "site": "bafy..locationCommitment"
+          }
+        },
+        "next": []
+      },
+    ],
+  },
+}
+```
+
+Legacy receipt:
+
+```json
+{ 
+  // "/": "bafy..work",
+  "iss": "did:web:node1.storage",
+  "aud": "did:web:web3.storage",
+  "ran": {
+    "/": "bafy..accept"
+  },
+  "out": {
+    "ok": {
+      // result is commitment that is produced as result of "bafy..accept"
+      "site": "bafy..locationCommitment"
+    }
+  },
+  "fx": [],
+}
+```
+
+4. Router fetches location commitment, and creates new commitment for the user.
+
+**Note: initially I thought this was a redelegation, but I think it's maybe not -- the commitment here is to a limited download URL of JUST the blob -- it's not a w3s gateway url with full download capabilities.
```

Comment:
I do think it should be re-delegation, which BTW can impose extra restrictions. I think it should be re-delegation because it would enable us to track reads.

I think gateway URLs are separate deal, specifically I think user should re-delegate location commitment to the gateway to authorize them to serve content through gateway because they will be billed for each read gateway will utilize. That also would enable user to revoke such delegation to stop gateway from serving said content at their expense.